### ISSUE TITLE
Memory leak when loop-referencing objects over contexts

### ISSUE THREAD

**Issue by d-ash** [nodejs/node-v0.x-archive#1155]
``` javascript
var vm = require('vm');

function run() {
  var obj = {}, ctx = vm.createContext({});

  obj.__ctx = ctx;
  ctx.__obj = obj;

  vm.Script.runInContext('var x = 1;', ctx);
}

for (var i = 0; i < 100000; i += 1) {
  run();
}
```

Leak occurs when runInContext() is executed.

**Comment by jasnell** [comment]
This may be a duplicate of #6552, which has a possible existing fix in io.js

**Comment by robraux** [comment]
@jasnell is a backport incoming for this issue?

**Comment by usmanhalalit** [comment]
I can reproduce the issue in v0.12.5 by using:

```
    vm.runInNewContext('var x = 1;', ctx);
```

Desperately looking for a fix. I wish I could help with a PR.

**Comment by robraux** [comment]
I can likewise reproduce it in node v0.12.7 and confirm that it does not exist in io.js 2.4.0. This is important enough to us that I'm happy to dig up where and how it was patches in io.js, along with patching the 0.12.x branch with a PR.

Before this, I'd like some confirmation from the corenode team that it's not a worthless endeavor.

**Comment by trevnorris** [comment]
That depends on what the patch entails. If the fix is from a V8 update then there's not much that can be done. First I'd try to replicate the bug with 1.0.0. If it still exists then it's more likely the fix can't be ported to v0.12.

**Comment by jasnell** [comment]
@robraux .. sorry, I'm a bit backlogged on github notifications and just now spotted your ping from a few days ago. I have not yet had the opportunity to dig into the details on this particular problem so I cannot say for sure, but the fix appears to have landed as part of the significant vm module updates that were made when the Contextify functionality was merged in. If that's the case, back porting may be a bit difficult but not impossible if a specific fix can be isolated and landed without introducing any API changes. Unfortunately, I won't have time to look at that myself for at least a couple of weeks so if you'd like to take a look, I'd happily review a PR :-)
